### ISSUE TITLE
Wave worker IOS 11.2

### ISSUE THREAD

**Issue by edubarbieri** [chris-rudmin/opus-recorder#128]
Hi,

I have mobile site with record audio feature.

In IOS 11.2.5 i have this error on waveWorker.min.js:

TypeError: undefined is not an object (evaluating 'this.recordedBuffers[0].length')

Any idea about this error?

**Comment by chris-rudmin** [comment]
@maxmatthews had the same issue when using the recorder in a react application. See #107.  I am not sure if he was able to resolve this problem. I might have some time later to investigate.

**Comment by chris-rudmin** [comment]
@edubarbieri I was able to validate the waveRecorder example works without issue on iOS 11.2.2 on iPhone 7.

**Comment by maxmatthews** [comment]
I was not able to resolve the problem. We ended up pushing back iOS web support and ended up using Cordova with native Siri recognition. We'll have the need for it in the future, so I'll definitely stay subscribed to this issue in case I can help out at all.

**Comment by maxmatthews** [comment]
@chris-rudmin 💯 that the example works, but converting that exact example in React seems to create issues (I think with the way the WebWorker works, but that's just a hypothesis)

**Comment by chris-rudmin** [comment]
@edubarbieri If you can provide me some sample code which replicates the issue, I will take a look at it.

**Comment by edubarbieri** [comment]
My code is similar to this:

```javascript

(function(){
    var recording = true;
    var audioRecorder = new Recorder({
        monitorGain : 0,
        numberOfChannels : 1,
        wavBitDepth : 16,
        encoderPath : '/js/waveWorker.min.js'   
    }); 
    audioRecorder.addEventListener('dataAvailable', processRecord);

    $('#record').click(function(event){
        event.preventDefault();
        if(!recording){
            startRecording();
        }else{
            stopRecord();
        }
    });

    function startRecording() {       
        audioRecorder.initStream().then(function(){
            audioRecorder.start();
            recording = true;
        });
        
    }

    function stopRecord(){
        audioRecorder && audioRecorder.stop();
        recording = false;
    }

    function processRecord(e){
        
        var dataBlob = new Blob([ e.detail ], {
            type : 'audio/wav'
        });

        //upload code
    }

})();

```

The function processRecord is not called.

I try change encoder to encoderWorker.min.js, in this case no error occurs, but the generated audio is empty.

**Comment by chris-rudmin** [comment]
Can you also specify the iOS 11.2.5 beta build # you are using to replicate
the issue?

**Comment by rowntreerob** [comment]
dont mean to intrude on the thread , but regarding the "audio is empty" issue, i think you can look [here](https://stackoverflow.com/questions/46363048/onaudioprocess-not-called-on-ios11/46534088#46534088)  for details on ios 11 issue requiring special co-ordination of UI event ( a click ) and the audioContext constructor ... on ios , it all has to happen in response to a click ( for most apps its just onStartRecord ( most apps dont do a separate 'init' step to allocate the context and the scriptNode. 

at risk of being off-topic, i use opus lib on ios and had to change the new audioContext call and the instance of the scriptNode to occur  on the callBack of the 'start' button ... This is just like the thread SO link above and would be just like this[ startRecord](https://github.com/GersonRosales/Record-Audios-and-Videos-with-getUserMedia/blob/master/example_3.html#L65) method from a much linked issue regarding webRTC and ios 11.   I could do a PR of my changes to 'recorder.js' - it does work on ios 11.2 but its kind of a mess?

**Comment by chris-rudmin** [comment]
@rowntreerob That's a solid theory. At a minimum we should add a note into the readme about that behaviour.

**Comment by rowntreerob** [comment]
https://gist.github.com/rowntreerob/614010bf487d28c4c9f9ddcd0f2d8e70 

i created a gist using a thead with a good description of the issue on ios and using a clone of the best , illustrative sample code i could find.

The comments at the top of the gist describe the 'audio record - no data' scenario where ios developers should start examining the code.... and what exactly they should verify

**Comment by chris-rudmin** [comment]
@edubarbieri @maxmatthews @rowntreerob Would love to have your feedback if these changes resolve your issues on iOS.

**Comment by chris-rudmin** [comment]
My changes do not work yet on iOS. I will let you know if I can get it to work.

**Comment by chris-rudmin** [comment]
I spent a long time debugging only to find iOS 11.2.2 and iOS 11.2.5 have a regression in their WebAssembly support.  https://bugs.webkit.org/show_bug.cgi?id=181781 :/

**Comment by chris-rudmin** [comment]
For the wave Recorder, this issue is resolved with the v3 release. The opus encoder will be resolved with the release of iOS 11.3
